src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c

```diff
@@ -0,0 +1,598 @@
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509ChainGetErrorCount(X509ChainContext* ctx)
+{
+    assert(ctx != NULL);
+    JNIEnv* env = GetJNIEnv();
+    int32_t count = (*env)->CallIntMethod(env, ctx->errorList, g_CollectionSize);
+    if (ctx->revocationErrorList != NULL)
+    {
+        count += (*env)->CallIntMethod(env, ctx->revocationErrorList, g_CollectionSize);
+    }
+
+    return count;
+}
+
+enum
+{
+    PKIXREASON_NAME_CHAINING,
+    PKIXREASON_INVALID_KEY_USAGE,
+    PKIXREASON_INVALID_POLICY,
+    PKIXREASON_NO_TRUST_ANCHOR,
+    PKIXREASON_UNRECOGNIZED_CRIT_EXT,
+    PKIXREASON_NOT_CA_CERT,
+    PKIXREASON_PATH_TOO_LONG,
+    PKIXREASON_INVALID_NAME,
+};
+
+enum
+{
+    BASICREASON_UNSPECIFIED,
+    BASICREASON_EXPIRED,
+    BASICREASON_NOT_YET_VALID,
+    BASICREASON_REVOKED,
+    BASICREASON_UNDETERMINED_REVOCATION_STATUS,
+    BASICREASON_INVALID_SIGNATURE,
+    BASICREASON_ALGORITHM_CONSTRAINED,
+};
+
+static PAL_X509ChainStatusFlags ChainStatusFromValidatorExceptionReason(JNIEnv* env, jobject reason)
+{
+    int value = (*env)->CallIntMethod(env, reason, g_EnumOrdinal);
+    if (g_CertPathExceptionBasicReasonClass != NULL
+        && (*env)->IsInstanceOf(env, reason, g_CertPathExceptionBasicReasonClass))
+    {
+        switch (value)
+        {
+            case BASICREASON_UNSPECIFIED:
+                return PAL_X509ChainPartialChain;
+            case BASICREASON_EXPIRED:
+            case BASICREASON_NOT_YET_VALID:
+                return PAL_X509ChainNotTimeValid;
+            case BASICREASON_REVOKED:
+                return PAL_X509ChainRevoked;
+            case BASICREASON_UNDETERMINED_REVOCATION_STATUS:
+                return PAL_X509ChainRevocationStatusUnknown;
+            case BASICREASON_INVALID_SIGNATURE:
+                return PAL_X509ChainNotSignatureValid;
+            case BASICREASON_ALGORITHM_CONSTRAINED:
+                return PAL_X509ChainPartialChain;
+        }
+    }
+    else if (g_PKIXReasonClass != NULL && (*env)->IsInstanceOf(env, reason, g_PKIXReasonClass))
+    {
+        switch (value)
+        {
+            case PKIXREASON_NAME_CHAINING:
+                return PAL_X509ChainPartialChain;
+            case PKIXREASON_INVALID_KEY_USAGE:
+                return PAL_X509ChainNotValidForUsage;
+            case PKIXREASON_INVALID_POLICY:
+                return PAL_X509ChainInvalidPolicyConstraints;
+            case PKIXREASON_NO_TRUST_ANCHOR:
+                return PAL_X509ChainPartialChain;
+            case PKIXREASON_UNRECOGNIZED_CRIT_EXT:
+                return PAL_X509ChainHasNotSupportedCriticalExtension;
+            case PKIXREASON_NOT_CA_CERT:
+                return PAL_X509ChainUntrustedRoot;
+            case PKIXREASON_PATH_TOO_LONG:
+                return PAL_X509ChainInvalidBasicConstraints;
+            case PKIXREASON_INVALID_NAME:
+                return PAL_X509ChainInvalidNameConstraints;
+        }
+    }
+
+    return PAL_X509ChainPartialChain;
```

Comment:
Can we do better than this?

The main thing is that we never return `true` from chain.Build if we don't know why something failed, so this is sort of OK, but I know I'm about to ask why a PartialChain is popping up and I'm guessing it's this fallback here.

On macOS our details processing comes from weak string processing, and anytime we get an unknown value we throw an exception (since we don't know if unknown means fatal or innocuous).

Reply:
A big source of `PartialChain` without any details would be from the initial build (instead of the second-pass validation). Unfortunately, all seem to come as a `CertPathBuilderException` with a generic message (`unable to find valid certification path to requested target`) and no wrapped exception, so I don't think there's any better we can do in that case.

In the case of the `CertPathValidatorException` without a reason (not set or on a version where it doesn't exist), I think we could do something like walk through any wrapped exceptions and check for specific exception types (e.g. the subclasses of [`CertificateException`](https://developer.android.com/reference/java/security/cert/CertificateException)) to determine a better status.

Testing on a version that has the reason API available, I was seeing errors with it set such that we could map them to a status, so I think it is on older versions where trying additional metrics would be helpful. I'll add it as an item to  #45741.